paddle/operators/detail/grpc_client.h

```diff
@@ -0,0 +1,269 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include <time.h>
+#include <chrono>
+#include <ctime>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "paddle/framework/data_type.h"
+#include "paddle/framework/lod_tensor.h"
+#include "paddle/framework/scope.h"
+#include "paddle/framework/selected_rows.h"
+#include "paddle/operators/detail/simple_block_queue.h"
+
+#include "paddle/operators/detail/send_recv.grpc.pb.h"
+#include "paddle/operators/detail/send_recv.pb.h"
+
+#include <grpc++/grpc++.h>
+#include <grpc/support/log.h>
+
+namespace paddle {
+namespace operators {
+namespace detail {
+
+struct VarHandle {
+  std::string ep;
+  std::string name;
+  std::string String() const {
+    std::ostringstream s;
+    s << "name:" << name << " ep:" << ep;
+    return s.str();
+  }
+};
+
+struct SendStatus {
+  std::string error;
+  std::chrono::system_clock::time_point start;
+  std::chrono::system_clock::time_point end;
+  VarHandle var;
+
+  std::string String() const {
+    std::time_t t0 = std::chrono::system_clock::to_time_t(start);
+    std::time_t t1 = std::chrono::system_clock::to_time_t(end);
+    std::ostringstream s;
+    s << var.String() << " start_time:" << std::ctime(&t0)
+      << " end_time:" << std::ctime(&t1);
+    return s.str();
+  }
+};
+
+class ClientBase {
```

Comment:
Would an RPCClient implementing `SendVariable` and `GetVariable` be sufficient?

members are the same and send, get should be called on the same RPC client stub
```
+  std::shared_ptr<grpc::CompletionQueue> cq_;
 +  std::shared_ptr<grpc::Channel> ch_;
 +  std::string ep_;
 +  std::unique_ptr<sendrecv::SendRecvService::Stub> stub_;
 +  std::unique_ptr<grpc::ClientContext> context_;
 +  std::unique_ptr<grpc::Status> status_;
 +  std::unique_ptr<sendrecv::VoidMessage> reply_;
 +  std::unique_ptr<grpc::ClientAsyncResponseReader<sendrecv::VoidMessage>> rpc_;
 +  SendStatus send_status_;
```

Reply:
Done.thanks.